### PR TITLE
[COLLECTIONS-795] Add a new Iterator to allowing zipping over two iterators of different types

### PR DESCRIPTION
**Pull request by anantdamle**
[COLLECTIONS-795] Add a new Iterator to allowing zipping over two iterators of different types.

### CODE REVIEW

File: src/main/java/org/apache/commons/collections4/iterators/ZippedTupleIterator.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+ *   while (zippedIterator.hasNext()) {
+ *     ZippedTuple<Integer, String> item = zippedIterator.next();
+ *     ...
+ *   }
+ * }</pre>
+ *
+ * @param <L> the left elements' type
+ * @param <R> the right elements' type
+ */
+public class ZippedTupleIterator<L, R> implements Iterator<ZippedTuple<L, R>> {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/iterators/ZippedTupleIterator.java:50]: Why is this prefixed with "zip" when it has nothing to do with zip files?

Comment (kinow) [reply src/main/java/org/apache/commons/collections4/iterators/ZippedTupleIterator.java:50]: I think it is related to [zipper](https://en.wikipedia.org/wiki/Zipper_(data_structure)). Like Python stdlib's [zip function](https://docs.python.org/3/library/functions.html#zip). Scala and Clojure also have a similar function I think.

File: src/test/java/org/apache/commons/collections4/iterators/ZippedTupleIteratorTest.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+      ZippedTuple<String, Integer> zippedItem = zipPairIterator.next();
+
+      assertEquals(
+          String.format("{%s, %s}", zippedItem.getLeftItem(), zippedItem.getRightItem()),
+          zippedItem.toString());
+    }
+
+    assertFalse(zipPairIterator.hasNext());
+  }
+}
```
Comment (kinow) [inline src/test/java/org/apache/commons/collections4/iterators/ZippedTupleIteratorTest.java:150]: Missing new line

File: src/main/java/org/apache/commons/collections4/iterators/ZippedTupleIterator.java (hunk 1 above)
Comment (garydgregory) [reply src/main/java/org/apache/commons/collections4/iterators/ZippedTupleIterator.java:50]: Maybe but the prefix here is "zipped" not "zipper", terrible name IMO.

Comment (kinow) [reply src/main/java/org/apache/commons/collections4/iterators/ZippedTupleIterator.java:50]: I'm not good naming things, so can't really suggest a better one. The code example in the Python docs uses `zipped` too for the variable that contains the tuples returned with `zip` (that were zipped), so that should sound OK for users coming from Python (or JS using Lodash's [zip/unzip](https://lodash.com/docs/#unzip) which I'm using in my current project). But if that's confusing for other users, we can try to think in a better name I think.

File: src/test/java/org/apache/commons/collections4/iterators/ZippedTupleIteratorTest.java (hunk 2 above)
Comment (anantdamle) [reply src/test/java/org/apache/commons/collections4/iterators/ZippedTupleIteratorTest.java]: Done/

File: src/main/java/org/apache/commons/collections4/iterators/ZippedTupleIterator.java (hunk 1 above)
Comment (anantdamle) [reply src/main/java/org/apache/commons/collections4/iterators/ZippedTupleIterator.java]: Updated to use `PairedIterator` and `PairedIterable` names. Hope is acceptable to all.

**Comment by XenoAmess** [comment]
> The CI flagged failing tests are not stemming from any code that I have touched,
> How to resolve?

checkstyles fails.

<details>
https://travis-ci.com/github/apache/commons-collections/jobs/508633787

[ERROR] src/main/java/org/apache/commons/collections4/IterableUtils.java:[574,21] (whitespace) WhitespaceAfter: ',' is not followed by whitespace.
[ERROR] src/main/java/org/apache/commons/collections4/IterableUtils.java:[574,46] (whitespace) WhitespaceAfter: ',' is not followed by whitespace.
[ERROR] src/main/java/org/apache/commons/collections4/IterableUtils.java:[578,47] (whitespace) WhitespaceAfter: ',' is not followed by whitespace.
[ERROR] src/main/java/org/apache/commons/collections4/IterableUtils.java:[580,41] (whitespace) WhitespaceAfter: ',' is not followed by whitespace.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[74,7] (indentation) Indentation: 'ctor def' child has incorrect indentation level 6, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[75,7] (indentation) Indentation: 'ctor def' child has incorrect indentation level 6, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[88,7] (indentation) Indentation: 'method def' child has incorrect indentation level 6, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[101,7] (indentation) Indentation: 'method def' child has incorrect indentation level 6, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[114,7] (indentation) Indentation: 'method def' child has incorrect indentation level 6, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[125,7] (indentation) Indentation: 'if' has incorrect indentation level 6, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[126,9] (indentation) Indentation: 'if' child has incorrect indentation level 8, expected level should be 12.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[127,7] (indentation) Indentation: 'if rcurly' has incorrect indentation level 6, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[129,7] (indentation) Indentation: 'method def' child has incorrect indentation level 6, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[140,5] (indentation) Indentation: 'member def modifier' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[142,5] (indentation) Indentation: 'member def modifier' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[144,5] (indentation) Indentation: 'ctor def modifier' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[145,7] (indentation) Indentation: 'ctor def' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[146,7] (indentation) Indentation: 'ctor def' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[147,5] (indentation) Indentation: 'ctor def rcurly' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[156,5] (indentation) Indentation: 'method def modifier' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[157,7] (indentation) Indentation: 'method def' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[158,5] (indentation) Indentation: 'method def rcurly' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[160,5] (indentation) Indentation: 'method def modifier' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[161,7] (indentation) Indentation: 'method def' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[162,5] (indentation) Indentation: 'method def rcurly' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[164,5] (indentation) Indentation: 'method def modifier' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[165,7] (indentation) Indentation: 'method def' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[166,5] (indentation) Indentation: 'method def rcurly' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[168,5] (indentation) Indentation: 'method def modifier' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[170,7] (indentation) Indentation: 'method def' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[171,5] (indentation) Indentation: 'method def rcurly' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:[172,3] (indentation) Indentation: 'class def rcurly' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/main/java/org/apache/commons/collections4/IteratorUtils.java:[921,21] (whitespace) WhitespaceAfter: ',' is not followed by whitespace.
[ERROR] src/main/java/org/apache/commons/collections4/IteratorUtils.java:[921,41] (whitespace) WhitespaceAfter: ',' is not followed by whitespace.
[ERROR] src/main/java/org/apache/commons/collections4/PairedIterable.java:[85,21] (whitespace) WhitespaceAfter: ',' is not followed by whitespace.
[ERROR] src/main/java/org/apache/commons/collections4/PairedIterable.java:[85,41] (whitespace) WhitespaceAfter: ',' is not followed by whitespace.
[ERROR] src/main/java/org/apache/commons/collections4/PairedIterable.java:[97,31] (whitespace) WhitespaceAfter: ',' is not followed by whitespace.
[ERROR] src/test/java/org/apache/commons/collections4/PairedIterableTest.java:[64,9] (whitespace) WhitespaceAfter: 'for' is not followed by whitespace.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[33,3] (indentation) Indentation: 'ctor def modifier' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[34,5] (indentation) Indentation: 'ctor def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[35,3] (indentation) Indentation: 'ctor def rcurly' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[39,3] (indentation) Indentation: 'member def modifier' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[40,3] (indentation) Indentation: 'member def modifier' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[41,3] (indentation) Indentation: 'member def modifier' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[42,3] (indentation) Indentation: 'member def modifier' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[45,3] (indentation) Indentation: 'member def modifier' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[46,3] (indentation) Indentation: 'member def modifier' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[48,3] (indentation) Indentation: 'method def modifier' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[50,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[52,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[53,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[54,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[55,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[57,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[59,5] (indentation) Indentation: 'for' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[60,7] (indentation) Indentation: 'for' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[61,7] (indentation) Indentation: 'for' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[62,5] (indentation) Indentation: 'for rcurly' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[64,5] (indentation) Indentation: 'for' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[65,7] (indentation) Indentation: 'for' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[66,7] (indentation) Indentation: 'for' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[67,5] (indentation) Indentation: 'for rcurly' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[68,3] (indentation) Indentation: 'method def rcurly' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[70,3] (indentation) Indentation: 'method def modifier' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[72,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[73,3] (indentation) Indentation: 'method def rcurly' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[77,3] (indentation) Indentation: 'method def modifier' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[79,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[80,3] (indentation) Indentation: 'method def rcurly' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[82,3] (indentation) Indentation: 'method def modifier' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[84,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[85,3] (indentation) Indentation: 'method def rcurly' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[87,3] (indentation) Indentation: 'method def modifier' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[88,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[92,5] (indentation) Indentation: 'for' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[93,7] (indentation) Indentation: 'for' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[94,7] (indentation) Indentation: 'for' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[96,7] (indentation) Indentation: 'for' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[96,44] (whitespace) NoWhitespaceBefore: ',' is preceded with whitespace.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[97,7] (indentation) Indentation: 'for' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[97,41] (whitespace) NoWhitespaceBefore: ',' is preceded with whitespace.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[98,5] (indentation) Indentation: 'for rcurly' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[100,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[101,3] (indentation) Indentation: 'method def rcurly' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[103,3] (indentation) Indentation: 'method def modifier' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[104,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[108,5] (indentation) Indentation: 'for' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[109,7] (indentation) Indentation: 'for' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[110,7] (indentation) Indentation: 'for' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[112,7] (indentation) Indentation: 'for' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[112,44] (whitespace) NoWhitespaceBefore: ',' is preceded with whitespace.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[113,7] (indentation) Indentation: 'for' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[113,41] (whitespace) NoWhitespaceBefore: ',' is preceded with whitespace.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[114,5] (indentation) Indentation: 'for rcurly' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[116,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[117,3] (indentation) Indentation: 'method def rcurly' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[119,3] (indentation) Indentation: 'method def modifier' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[120,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[123,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[124,3] (indentation) Indentation: 'method def rcurly' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[126,3] (indentation) Indentation: 'method def modifier' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[127,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[130,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[131,3] (indentation) Indentation: 'method def rcurly' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[134,3] (indentation) Indentation: 'method def modifier' has incorrect indentation level 2, expected level should be 4.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[135,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[139,5] (indentation) Indentation: 'for' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[140,7] (indentation) Indentation: 'for' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[141,7] (indentation) Indentation: 'for' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[143,7] (indentation) Indentation: 'for' child has incorrect indentation level 6, expected level should be 12.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[146,5] (indentation) Indentation: 'for rcurly' has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[148,5] (indentation) Indentation: 'method def' child has incorrect indentation level 4, expected level should be 8.
[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[149,3] (indentation) Indentation: 'method def rcurly' has incorrect indentation level 2, expected level should be 4.
</details>

**Comment by kinow** [comment]
> The CI flagged failing tests are not stemming from any code that I have touched,
> How to resolve?

Click on the link of the failed Travis CI builds. Then choose one of the JDK versions (e.g. 8). There you should find the logs of the job that failed. [For instance](https://travis-ci.com/github/apache/commons-collections/jobs/508633788):

>[ERROR] src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:[149,3] (indentation) Indentation: 'method def rcurly' has incorrect indentation level 2, expected level should be 4.

**Comment by anantdamle** [comment]
thanks @XenoAmess and @kinow 
Fixed checkstyles issues (also checked locally using `mvn checkstyle:check`)

The sytle checks & corrections can be simplified using `spotless` plugin.
Should I create a separate PR to implement spotless plugin for auto style correction?

**Comment by anantdamle** [comment]
Added tests to improve coverage

**Comment by anantdamle** [comment]
What are the next steps?

File: src/main/java/org/apache/commons/collections4/IteratorUtils.java [hunk 3]
```diff
@@ -907,6 +908,20 @@ public static NodeListIterator nodeListIterator(final Node node) {
...
 
+    /**
+     * Returns an iterator that provides the elements contained in a pair of Iterators.
+     *
+     * @param <L> the left elements' type
+     * @param <R> the right elements' type
+     * @param left the iterator for the left side elements
+     * @param right the iterator for the right side elements
+     * @return an iterator, to iterate over the decorated iterators together until one is exhausted
+     * @throws NullPointerException if any iterator is null
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/IteratorUtils.java:919]: All new public and protected methods should have a Javadoc since 4.5 tag.

File: src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+import java.util.UUID;
+import org.apache.commons.collections4.IteratorUtils;
+import org.apache.commons.collections4.iterators.PairedIterator.PairedItem;
+
+
+/** Unit test suite for {@link ZippingIterator}. */
+public final class PairedIteratorTest
+    extends AbstractIteratorTest<PairedItem<String, Integer>> {
+
+    //------------------------------------------------------------ Conventional
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:31]: No need for these inline comments, they document nothing.

File: src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+        return PairedItem.of(leftIterator.next(), rightIterator.next());
+    }
+
+    /**
+     * An immutable tuple class to represent elements from both the iterators.
+     *
+     * @param <L> the left elements' type
+     * @param <R> the right elements' type
+     */
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/iterators/PairedIterator.java:135]: We need to think about this because Common Lang already defines different kinds of Pair classes...

File: src/test/java/org/apache/commons/collections4/PairedIterableTest.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.Iterator;
+import java.util.Random;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import org.apache.commons.collections4.iterators.PairedIterator.PairedItem;
+import org.junit.Before;
+import org.junit.Test;
+
+public final class PairedIterableTest {
+    private ArrayList<String> smallStringsList = null;
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/collections4/PairedIterableTest.java:31]: Don't override default vslues with default values.

File: src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+    //------------------------------------------------------------ Lifecycle
+
+    private ArrayList<String> smallStringsList = null;
+    private ArrayList<String> largeStringsList = null;
+    private ArrayList<Integer> smallIntsList = null;
+    private ArrayList<Integer> largeIntsList = null;
+
+    // Unequal sized lists
+    private static final int SMALL_LIST_SIZE = 20;
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:45]: If you expect the sizes to be different, you should assert that fact somewhere.

**Review by garydgregory — CHANGES_REQUESTED** [review]
I'll try to take a look again later this week but did I miss something or are there no tests for empty lists, lists of unequal sizes, null inputs?

File: src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java (hunk 7 above)
Comment (anantdamle) [reply src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:45]: Its asserted in two tests: `testLeftIteratorLargerThanRight` and `testRightIteratorLargerThanLeft`

File: src/main/java/org/apache/commons/collections4/IteratorUtils.java (hunk 3 above)
Comment (anantdamle) [reply src/main/java/org/apache/commons/collections4/IteratorUtils.java:919]: Added a more descriptive Javadoc.

File: src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java (hunk 4 above)
Comment (anantdamle) [reply src/test/java/org/apache/commons/collections4/iterators/PairedIteratorTest.java:31]: Removed.
Had copied them from another test.

File: src/test/java/org/apache/commons/collections4/PairedIterableTest.java (hunk 6 above)
Comment (anantdamle) [reply src/test/java/org/apache/commons/collections4/PairedIterableTest.java:31]: done.

File: src/main/java/org/apache/commons/collections4/IteratorUtils.java [hunk 8]
```diff
@@ -127,7 +87,7 @@ private IteratorUtils() {}
      * @return an iterator over nothing
      */
     public static <E> ResettableIterator<E> emptyIterator() {
-        return EmptyIterator.<E>resettableEmptyIterator();
+        return EmptyIterator.resettableEmptyIterator();
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/IteratorUtils.java:90]: These changes are unrelated, small PRs are better for reviewers ;-)

File: src/main/java/org/apache/commons/collections4/IterableUtils.java [hunk 9]
```diff
@@ -554,6 +555,34 @@
         };
     }
 
+    /**
+     * Provides iteration over the elements contained in a pair of Iterables in-tandem.
+     * <p>
+     * The returned iterable has an iterator that traverses the elements in {@code a}
+     * and {@code b} together until one of the iterables is traversed completely.
+     * <p>
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/IterableUtils.java:563]: Close HTML tags.

File: src/main/java/org/apache/commons/collections4/IterableUtils.java [hunk 10]
```diff
@@ -554,6 +555,34 @@
...
+     * The returned iterable's iterator does NOT support {@code remove()}.
+     *
+     * @param <L> the left elements' type
+     * @param <R> the right elements' type
+     * @param left the iterable for the left side elements
+     * @param right the iterable for the right side elements
+     * @return an iterable, over the decorated iterables to traverse them together until one is
+     * exhausted
+     * @throws NullPointerException if any iterator is null
+     */
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/IterableUtils.java:573]: New public and protected method should have `@since 4.5` Javadoc tags.

**Review by garydgregory — CHANGES_REQUESTED** [review]
Hi All,
See 2 minor comments. Another class to inspect is the existing `ZippingIterator` which should be considered for reuse; but note its different 'switching' aspect. Also, `ZippingIterator` will work for any number of input iterators (`ZippingIterator.ZippingIterator(Iterator<? extends E>...)`), not just two. So perhaps this PR should as well, which would remove the need for the pair class.

**Comment by anantdamle** [comment]
@garydgregory there is one big difference with ZippingIterator, ZippingIterator allows iterating over iteratord of same generic type, whereas the proposed solution allows using two different types and hence needs a Pair equivalent class.

**Comment by anantdamle** [comment]
@garydgregory gentle bump up.

File: src/main/java/org/apache/commons/collections4/IteratorUtils.java (hunk 8 above)
Comment (anantdamle) [reply src/main/java/org/apache/commons/collections4/IteratorUtils.java:90]: Done.

File: src/main/java/org/apache/commons/collections4/IterableUtils.java (hunk 9 above)
Comment (anantdamle) [reply src/main/java/org/apache/commons/collections4/IterableUtils.java:563]: Done.

File: src/main/java/org/apache/commons/collections4/IterableUtils.java (hunk 10 above)
Comment (anantdamle) [reply src/main/java/org/apache/commons/collections4/IterableUtils.java:573]: Done

**Comment by anantdamle** [comment]
Gentle bump: any comments or changes today I should be doing?

**Comment by anantdamle** [comment]
Gentle Bump: Any review comments or improvements

**Comment by garydgregory** [comment]
No bumping needed. We are all volunteers. We are all busy. Please be patient.

**Comment by anantdamle** [comment]
sorry to nag! Any updates?

**Comment by garydgregory** [comment]
@anantdamle 
We are volunteers here, we do what we can when we can ;-)

**Comment by garydgregory** [comment]
I'll try to take a look this week...

**Comment by anantdamle** [comment]
Any updates on this one?
